### ISSUE TITLE
Pickling fails with local functions (e.g. ee.Image.expression())

### ISSUE THREAD

**Issue by davemlz** [aazuspan/eexarray#3]
Hi, @aazuspan!

First of all, WOW! Your work with `eexarray` is amazing, keep it going! :rocket: 

I was using your dev repo to try to convert a S2 collection to xarray, and it works, but, when I compute a spectral index using `eemont` (that uses `ee.Image.expression`) it doesn't work:

**This works!**

```python
import ee, eemont, eexarray

ee.Initialize()

tw = ee.Geometry.Point([10.4522,51.0792])
bf = tw.buffer(500)
xt = bf.bounds()

S2 = ee.ImageCollection("COPERNICUS/S2_SR") \
    .filterBounds(xt) \
    .preprocess() \
    .map(lambda x: x.addBands(x.normalizedDifference(["B8","B4"]).rename("NDVI"))) \
    .limit(10) \
    .map(lambda x: x.clip(xt)) \
    .eex.resample_daily(reducer = ee.Reducer.median())

S2eex = S2.eex.to_xarray(scale=10) 
```

**This doesn't work (using eemont)**

```python
import ee, eemont, eexarray

ee.Initialize()

tw = ee.Geometry.Point([10.4522,51.0792])
bf = tw.buffer(500)
xt = bf.bounds()

S2 = ee.ImageCollection("COPERNICUS/S2_SR") \
    .filterBounds(xt) \
    .preprocess() \
    .spectralIndices("NDVI") \
    .limit(10) \
    .map(lambda x: x.clip(xt)) \
    .eex.resample_daily(reducer = ee.Reducer.median())

S2eex = S2.eex.to_xarray(scale=10) 
```

**This doesn't work (not using eemont)**

```python
import ee, eemont, eexarray

ee.Initialize()

tw = ee.Geometry.Point([10.4522,51.0792])
bf = tw.buffer(500)
xt = bf.bounds()

def addExpressionNDVI(x):
    params = {"N": x.select("B8"),"R": x.select("B4")}
    NDVI = x.expression("(N-R)/(N+R)",params).rename("NDVI")
    return x.addBands(NDVI)

S2 = ee.ImageCollection("COPERNICUS/S2_SR") \
    .filterBounds(xt) \
    .preprocess() \
    .map(addExpressionNDVI) \
    .limit(10) \
    .map(lambda x: x.clip(xt)) \
    .eex.resample_daily(reducer = ee.Reducer.median())

S2eex = S2.eex.to_xarray(scale=10) 
```

**Error**

```python
AttributeError: Can't pickle local object 'Image.expression.<locals>.ReinterpretedFunction'
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-37-94ef9caa673d> in <module>
----> 1 S2eex = S2.eex.to_xarray(scale=10)

~/anaconda3/envs/gee/lib/python3.9/site-packages/eexarray/ImageCollection.py in to_xarray(self, path, region, scale, crs, masked, nodata, num_cores, progress, max_attempts)
     90             collection = self._rename_by_time()
     91 
---> 92             files = collection.eex.to_tif(
     93                 out_dir=tmp,
     94                 region=region,

~/anaconda3/envs/gee/lib/python3.9/site-packages/eexarray/ImageCollection.py in to_tif(self, out_dir, prefix, region, scale, crs, file_per_band, masked, nodata, num_cores, progress, max_attempts)
    198                 max_attempts=max_attempts,
    199             )
--> 200             tifs = list(
    201                 tqdm(
    202                     p.imap(params, imgs),

~/anaconda3/envs/gee/lib/python3.9/site-packages/tqdm/std.py in __iter__(self)
   1183 
   1184         try:
-> 1185             for obj in iterable:
   1186                 yield obj
   1187                 # Update and possibly print the progressbar.

~/anaconda3/envs/gee/lib/python3.9/multiprocessing/pool.py in next(self, timeout)
    868         if success:
    869             return value
--> 870         raise value
    871 
    872     __next__ = next                    # XXX

~/anaconda3/envs/gee/lib/python3.9/multiprocessing/pool.py in _handle_tasks(taskqueue, put, outqueue, pool, cache)
    535                         break
    536                     try:
--> 537                         put(task)
    538                     except Exception as e:
    539                         job, idx = task[:2]

~/anaconda3/envs/gee/lib/python3.9/multiprocessing/connection.py in send(self, obj)
    209         self._check_closed()
    210         self._check_writable()
--> 211         self._send_bytes(_ForkingPickler.dumps(obj))
    212 
    213     def recv_bytes(self, maxlength=None):

~/anaconda3/envs/gee/lib/python3.9/multiprocessing/reduction.py in dumps(cls, obj, protocol)
     49     def dumps(cls, obj, protocol=None):
     50         buf = io.BytesIO()
---> 51         cls(buf, protocol).dump(obj)
     52         return buf.getbuffer()
     53 

AttributeError: Can't pickle local object 'Image.expression.<locals>.ReinterpretedFunction'
```

**Versions**

- xarray 0.19.0
- earthengine-api 0.1.277
- eemont 0.2.5
- python 3.9

It seems to be something related specifically to that earthengine-api method, but if you can find a workaround, that would be amazing! :rocket: 

And again, thank you very much for `eexarray`!

**Comment by aazuspan** [comment]
Hey @davemlz, thanks for pointing this out! I'm glad you're finding it useful!!

If you want to update your install to the latest commit, I just added a quick workaround by setting `num_cores=1`. Unfortunately that will deactivate parallel processing and slow down your downloads, but it should bypass the error.

```python
import ee, eemont, eexarray

ee.Initialize()

tw = ee.Geometry.Point([10.4522,51.0792])
bf = tw.buffer(500)
xt = bf.bounds()

S2 = ee.ImageCollection("COPERNICUS/S2_SR") \
    .filterBounds(xt) \
    .preprocess() \
    .spectralIndices("NDVI") \
    .limit(10) \
    .map(lambda x: x.clip(xt)) \
    .eex.resample_daily(reducer = ee.Reducer.median())

S2eex = S2.eex.to_xarray(scale=10, num_cores=1) 
```

The issue is that `eexarray` uses `multiprocessing` to parallelize downloads, and `multiprocessing` annoyingly can't handle local functions (in this case, `Image.expression.<locals>.ReinterpretedFunction`) because they can't be pickled. I was hoping that wouldn't cause any problems for users, but it looks like I need to find another solution! 

I'll keep this open and let you know when I've got a permanent fix!
